enhancements/network/service-egress-traffic-steering.md

```diff
@@ -0,0 +1,284 @@
+Name:                     some-service
+Namespace:                default
+Annotations:              k8s.ovn.org/egress-service: '{"network-hint":172.19.0.0/24}' (set by user)
+                          k8s.ovn.org/egress-service-host: "ovn-worker"                (set by ovn-k)
+Type:                     LoadBalancer
+LoadBalancer Ingress:     172.19.0.100
+Endpoints:                10.244.0.3:8080,10.244.1.6:8080
+```
+the egress traffic flow for the pod `10.244.1.6` on `ovn-worker2` towards an external destination (172.19.0.5) will look like:
+```none
+                           ┌────────────────────┐
+                           │                    │
+                           │external destination│
+                           │    172.19.0.5      │
+                           │                    │
+                           └───▲────────────────┘
+                               │
+           5. packet reaches   │                      2. router policy rereoutes it
+              the external     │                         to ovn-worker's mgmt port
+              destination with │                      ┌──────────────────┐
+              src ip:          │                  ┌───┤ovn cluster router│
+              172.19.0.100     │                  │   └───────────▲──────┘
+                               │                  │               │
+                               │                  │               │1. packet to 172.19.0.5
+                            ┌──┴───┐        ┌─────▼┐              │   heads to the cluster router
+                         ┌──┘ eth1 └──┐  ┌──┘ mgmt └──┐           │   as usual
+                         │ 172.19.0.2 │  │ 10.244.0.2 │           │
+                         ├─────▲──────┴──┴─────┬──────┤           │   ┌────────────────┐
+      4. an iptables rule│     │   ovn-worker  │3.    │           │   │  ovn-worker2   │
+         that SNATs to   │     │               │      │           │   │                │
+         the service's ip│     │               │      │           │   │                │
+         and sends it    │     │               │      │           │   │ ┌────────────┐ │
+         through eth1    │     │4.┌────────┐   │      │           └───┼─┤    pod     │ │
+         is hit          │     └──┤iptables◄───┘      │               │ │ 10.244.1.6 │ │
+                         │        └────────┘          │               │ └────────────┘ │
+                         │                            │               │                │
+                         └────────────────────────────┘               └────────────────┘
+                          3. from the mgmt port it goes to
+                             ovn-worker's iptables rules
+                             before heading out
+
+```
+As mentioned earlier, for the opposite direction (ingress/external client initiates) to work properly the LoadBalancer provider needs to announce the service only from `ovn-worker`.
+
+### Usage Example
```

Comment:
This is quite complicated... is there any reason OVN-Kubernetes or CNO couldn't create the MetalLB objects based on the annotations on the service?

Reply:
yes, for that you'll probably need to replicate the fields of the MetalLB resources in that annotation in order to fully control them, which would be messy? not counting for whatever complicated configuration a user might want for advertising the service.
another thing is that I think we shouldn't force a user (upstream wise) to use MetalLB as their LoadBalancer provider - ideally we should not care about the ingress part from our side, and let the user handle it however they want with MetalLB or whatever LoadBalancer that supports node selectors (in theory they could set static routes on their external hosts to the right node and this solution would work).

Reply:
I think adding the capability to CNO is a good idea as it's OCP only. not sure if adding metallb resources to OVN-Kubernetes will be possible

Reply:
We had this discussion originally, that would mean (possibly) duplicating all the bgp advertisement (assuming we care about bgp only, l2 is lighter anyway) fields in https://metallb.universe.tf/apis/#metallb.io/v1beta1.BGPAdvertisement (minus the pool selector I guess) in another resource, plus having to keep them in sync (in terms of fields).
Same for IPAddressPool's fields.